### ISSUE TITLE
CUDA Memory Optimization and Data Period Refinement

### ISSUE THREAD

**Issue by alexoms** [OptimalMatch/temporal-trading-agents#65]
## Summary
Series of optimizations to fix CUDA out-of-memory errors and improve prediction quality by refining data period selection for different intervals.

### 1. CUDA OOM on Daily Analysis (Initial)
**Issue**: Training with 5y of daily data caused CUDA OOM
- 1823 data points with batch_size=1280 exceeded GPU memory
- Error: "Tried to allocate 90.00 MiB... 99.5% utilization"

### 2. Wildly Pessimistic Predictions  
**Issue**: Daily predictions showing -11% to -26% drops (BTC from $102k → $75k-$90k)
- 5y of daily data goes back to when BTC was $20k-$30k
- Models learned mean reversion to old price levels
- Not predictive for current market conditions

### 3. Training Dataset Size Error
**Issue**: `__len__() should return >= 0` error on 21-day forecasts
- 2y data (~729 samples) + 90-day lookback + 21-day forecast
- Validation set: ~102 samples, Required: 111 samples
- Math doesn't work: 102 - 111 = -9

### 4. CUDA OOM on Parallel Training
**Issue**: Multiple strategies running simultaneously caused OOM
- 4 processes × 5-6 GB each = 23.45/23.54 GB (99.6%)
- Left no headroom for memory spikes

### Adaptive Batch Sizing (Commit 7ffbb9e)
```python
if interval == '1d':
    batch_size = 512  # Daily: fewer samples
else:
    batch_size = 1280  # Hourly: more samples
```

### PyTorch Memory Optimization (Commit 15cc33b)
- Added `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`
- Reduced daily batch_size: 512 → 256
- Reduced hourly batch_size: 1280 → 1024
- Reduces memory fragmentation

### Interval-Aware Period Preference (Commit 40ed288)
```python
if interval == '1d':
    preferred_periods = ['2y', '1y', '6mo', '5y']  # Prefer recent data
else:
    preferred_periods = ['5y', '2y', '1y', '6mo']  # Prefer more history
```
**Impact**: Daily analysis now uses 2y (~729 samples) instead of 5y (~1823)
**Result**: Predictions now -7% to +2% range (realistic vs. -11% to -26%)

### Dynamic Lookback Adjustment (Commit 6071d61)
```python
if period in ['2y', '1y', '6mo'] and lookback > 60:
    lookback = 60
    print(f"⚠️  Reduced {config['name']} lookback to {lookback} days")
```
Ensures validation sets have enough samples for all forecast horizons.

### Aggressive Memory Conservation (Commit b73a580)
- Daily batch_size: 256 → 128
- Hourly batch_size: 1024 → 768  
- Added `torch.cuda.empty_cache()` after each model trains
- Handles 4+ simultaneous training processes

### ✅ CUDA Memory Usage
**Before**: 23.43/23.54 GB (99.5% - unstable)
**After**: ~18/23.54 GB (76% - stable with headroom)

### ✅ Prediction Quality
**Before** (5y daily):
```
Day 7: -11% to -26% (pessimistic)
Predicting $75k-$90k
```
**After** (2y daily):
```
Day 7: -7% to +2% (realistic)
Predicting $95k-$103k
Current price: $102k
```

### ✅ Training Success Rate
**Before**: 21-day forecasts failing with dataset size errors
**After**: All horizons (3d, 7d, 14d, 21d) training successfully

### ✅ Parallel Training
**Before**: CUDA OOM when running multiple analyses
**After**: Multiple analyses running concurrently without errors

### Files Modified
- `examples/crypto_ensemble_forecast.py`: Adaptive batching, GPU cleanup
- `strategies/strategy_utils.py`: Period preferences, lookback limits
- `docker-compose.yml`: PyTorch memory optimization env var

### Batch Size Evolution
| Interval | Initial | After OOM #1 | After OOM #2 | Final |
|----------|---------|--------------|--------------|-------|
| Daily    | 1280    | 512          | 256          | 128   |
| Hourly   | 1280    | 1280         | 1024         | 768   |

### Memory Optimization Strategy
1. Interval-aware batch sizing
2. Dynamic lookback adjustment based on available data
3. Explicit GPU cache clearing between models
4. PyTorch expandable_segments for fragmentation reduction

## Testing
- [x] Single analysis completes without errors
- [x] Multiple parallel analyses complete without CUDA OOM
- [x] Predictions realistic for daily intervals
- [x] Fine-tuning working correctly (observed in logs)
- [x] All forecast horizons training successfully

## Related Work
Builds on auto-detection feature from PR #63 / Issue #64

## Commits
- 7ffbb9e Use adaptive batch size based on data interval
- 15cc33b Reduce batch sizes and enable PyTorch memory optimization
- 40ed288 Prefer shorter periods for daily intervals to avoid stale data
- 6071d61 Cap lookback at 60 days for shorter data periods
- b73a580 Reduce batch sizes and add GPU cache clearing for parallel training
